working-groups/pipelining/NOTES.md

```diff
@@ -0,0 +1,166 @@
+# pipelining Meeting Notes
+
+## Video 2019-04-05
+
+@alexcrichton and @nnethercote met on video for ~30m and talked about various
+aspects of implementing pipelining in the compiler.
+
+#### What are metadata/rlibs?
+
+First we talked a bit about what are rlibs/metadata files and how is this all
+going to be put together. The recap is:
+
+* Rustc can produce metadata files (`--emit metadata`). These metadata files are
+  like header files for Rust crates. They're internally a compiler-specific
+  binary format and cannot be inspected.
+
+* Rustc can also produce rlibs (`--emit link --crate-type lib`). An rlib is an
+  archive (a `*.a` file) which contains three things:
+
+  * Object code (`*.o`)
+  * Compressed bytecode (`*.bc.z`)
+  * Metadata (`metadata.bin`)
+
+  The metadata included here is the same as `--emit metadata`
+
+When you type `cargo build`, Cargo is likely building an `rlib` for almost all
+library crates in use. When you type `cargo check` Cargo produces metadata files
+for all crates.
+
+#### How are we going to pipeline?
+
+The goal of pipelining is similar to CPU instruction pipelining, which is to
+fill up available hardware with as much work as possible. This can increase
+overall throughput without actually speeding up the intermediate operations. For
+example let's say your compile looks like this:
+
+
+```
+            meta                  meta
+[ libA       |     ][ libB         |     ][ binary              ]
+1s          5s     10s            15s   20s                    30s
```

Comment:
Having a diagram is great, but it could be better, viz:
```
         meta                meta
[-libA----|--------][-libB----|--------][-binary-----------]
0s        5s       10s       15s       20s                30s
```
- Start as 0s, not 1s
- Better spacing of the mid-crate points
- Use '-' char to avoid big spaces that are hard to read
Similar changes are possible in the diagrams below.